apps/emqx/src/emqx_persistent_session_ds_router.erl

```diff
@@ -0,0 +1,230 @@
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_persistent_session_ds_router).
+
+-include("emqx.hrl").
+-include("emqx_persistent_session_ds/emqx_ps_ds_int.hrl").
+
+-export([init_tables/0]).
+
+%% Route APIs
+-export([
+    do_add_route/2,
+    do_delete_route/2,
+    match_routes/1,
+    lookup_routes/1,
+    foldr_routes/2,
+    foldl_routes/2
+]).
+
+-export([cleanup_routes/1]).
+-export([print_routes/1]).
+-export([topics/0]).
+
+-ifdef(TEST).
+-export([has_route/2]).
+-endif.
+
+-type dest() :: emqx_ds:session_id().
+
+-export_type([dest/0]).
+
+%%--------------------------------------------------------------------
+%% Table Initialization
+%%--------------------------------------------------------------------
+
+init_tables() ->
+    mria_config:set_dirty_shard(?PS_ROUTER_SHARD, true),
+    ok = mria:create_table(?PS_ROUTER_TAB, [
+        {type, bag},
+        {rlog_shard, ?PS_ROUTER_SHARD},
+        {storage, ram_copies},
+        {record_name, ps_route},
+        {attributes, record_info(fields, ps_route)},
+        {storage_properties, [
+            {ets, [
+                {read_concurrency, true},
+                {write_concurrency, true}
+            ]}
+        ]}
+    ]),
+    ok = mria:create_table(?PS_FILTERS_TAB, [
+        {type, ordered_set},
+        {rlog_shard, ?PS_ROUTER_SHARD},
+        {storage, ram_copies},
+        {record_name, ps_routeidx},
+        {attributes, record_info(fields, ps_routeidx)},
+        {storage_properties, [
+            {ets, [
+                {read_concurrency, true},
+                {write_concurrency, auto}
+            ]}
+        ]}
+    ]),
+    ok = mria:wait_for_tables([?PS_ROUTER_TAB, ?PS_FILTERS_TAB]),
+    ok.
+
+%%--------------------------------------------------------------------
+%% Route APIs
+%%--------------------------------------------------------------------
+
+-spec do_add_route(emqx_types:topic(), dest()) -> ok | {error, term()}.
+do_add_route(Topic, Dest) when is_binary(Topic) ->
+    case has_route(Topic, Dest) of
+        true ->
+            ok;
+        false ->
+            mria_insert_route(Topic, Dest)
+    end.
+
+-spec do_delete_route(emqx_types:topic(), dest()) -> ok | {error, term()}.
+do_delete_route(Topic, Dest) ->
+    case emqx_trie_search:filter(Topic) of
+        Words when is_list(Words) ->
+            K = emqx_topic_index:make_key(Words, Dest),
+            mria:dirty_delete(?PS_FILTERS_TAB, K);
+        false ->
+            mria_route_tab_delete(#ps_route{topic = Topic, dest = Dest})
+    end.
+
+%% @doc Take a real topic (not filter) as input, return the matching topics and topic
+%% filters associated with route destination.
+-spec match_routes(emqx_types:topic()) -> [emqx_types:route()].
+match_routes(Topic) when is_binary(Topic) ->
+    lookup_route_tab(Topic) ++
+        [match_to_route(M) || M <- match_filters(Topic)].
```

Comment:
There's potential for an optimization: currently, we don't need the list of matching routes, only the fact that there's at least one, and there's `emqx_topic_index:match/2` for that.